### PR TITLE
Language feature: range float types

### PR DESCRIPTION
**Pull request by cooldome**
For programs that are using floats rather heavily, it is very beneficial to be able to enforce the correctness of values and ranges is one of them. The most of infrastructure was already available in the compiler so it didn't require a lot of effort to implement (read: easy to maintain).

The trickiest part was to make `firstOrdValue` and `lastOrdValue` to works with floats and not to duplicate large parts of the code. Generic proc that return BiggestInt or BiggestFloat seems to work quite well.

I have updated the doc, changelog and add a test

### CODE REVIEW

**Comment by notTito** [comment]
I like range types. But the thing is the checks are completely removed in release mode. So how can you write code that depends on them and expect it to work? Especially in numerical computations.

**Comment by Araq** [comment]
> I like range types. But the thing is the checks are completely removed in release mode. So how can you write code that depends on them and expect it to work? Especially in numerical computations.

By testing your code?

**Comment by Araq** [comment]
@cooldome Please rebase so that we have decent test results.

File: compiler/types.nim [hunk 1]
```diff
@@ -587,79 +606,106 @@ proc typeToString(typ: PType, prefer: TPreferedDesc = preferName): string =
...
-    result = lastOrd(lastSon(t))
-  of tyProxy: result = 0
-  of tyOrdinal:
-    if t.len > 0: result = lastOrd(lastSon(t))
-    else: internalError("invalid kind for last(" & $t.kind & ')')
+    result = getValue[T](t.n.sons[1])
   else:
-    internalError("invalid kind for last(" & $t.kind & ')')
-    result = 0
+    when T is BiggestInt:
```
Comment (Araq) [inline compiler/types.nim]: I dislike this style. It's not a "generic", it's 2-3 different implementations which happen to use the same name. And they are not generic.

File: compiler/types.nim [hunk 2]
```diff
@@ -587,79 +606,106 @@ proc typeToString(typ: PType, prefer: TPreferedDesc = preferName): string =
...
+        case t.kind:
+          of tyFloat..tyFloat128: result = -Inf
+          else:
+            internalError("invalid kind for first(" & $t.kind & ')')
+            result = NaN
+
+proc firstOrd*(t: PType): BiggestInt {.inline.} =
+  firstValue[BiggestInt](t)
+
+proc lastValue*[T:BiggestInt|BiggestFloat](t: PType; fixedUnsigned = false): T =
```
Comment (Araq) [inline compiler/types.nim]: I dislike this style. It's not a "generic", it's 2-3 different implementations which happen to use the same name. And they are not generic.

File: compiler/sigmatch.nim [hunk 3]
```diff
@@ -633,20 +633,27 @@ proc procTypeRel(c: var TCandidate, f, a: PType): TTypeRelation =
...
+      # X..Y and C..D overlap iff (X <= D and C <= Y)
+      result = isConvertible
+    else:
+      result = isNone
+  
+  if f.isOrdinalType: 
+    check_range_in(BiggestInt)
   else:
-    result = isNone
+    check_range_in(BiggestFloat)
```
Comment (Araq) [inline compiler/sigmatch.nim]: Instead pass ``firstFloatValue, lastFloatValue`` to ``change_range_in``.

**Comment by Araq** [comment]
Also: Do not use underscores.

File: compiler/types.nim (hunk 1 above)
Comment (cooldome) [reply compiler/types.nim]: done, Introduced separate firstFloat and lastFloat procs

File: compiler/types.nim (hunk 2 above)
Comment (cooldome) [reply compiler/types.nim]: done

File: compiler/sigmatch.nim (hunk 3 above)
Comment (cooldome) [reply compiler/sigmatch.nim]: done

**Comment by cooldome** [comment]
Fixed all comments. Ready for the next round of review

File: doc/manual.rst [hunk 4]
```diff
@@ -741,22 +741,26 @@ For further details, see `Convertible relation
 
 Subrange types
 --------------
-A subrange type is a range of values from an ordinal type (the base
+A subrange type is a range of values from an ordinal or float point type (the base
```
Comment (Araq) [inline doc/manual.rst]: typo: "floating point"

File: compiler/sigmatch.nim [hunk 5]
```diff
@@ -633,20 +633,22 @@ proc procTypeRel(c: var TCandidate, f, a: PType): TTypeRelation =
...
-    result = isEqual
-  elif a0 >= f0 and a1 <= f1:
-    result = isConvertible
-  elif a0 <= f1 and f0 <= a1:
-    # X..Y and C..D overlap iff (X <= D and C <= Y)
-    result = isConvertible
-  else:
-    result = isNone
+  template checkRange[T](afirst, alast, ffirst, flast: T): TTypeRelation = 
+    if afirst == ffirst and alast == flast:
```
Comment (Araq) [inline compiler/sigmatch.nim]: Nitpick: I find the old names ``a0, ...`` easier to read.

File: compiler/sigmatch.nim [hunk 6]
```diff
@@ -633,20 +633,22 @@ proc procTypeRel(c: var TCandidate, f, a: PType): TTypeRelation =
...
+      isEqual
+    elif afirst >= ffirst and alast <= flast:
+      isConvertible
+    elif afirst <= flast and ffirst <= a_last:
```
Comment (Araq) [inline compiler/sigmatch.nim]: ``alast``

File: compiler/sigmatch.nim [hunk 7]
```diff
@@ -633,20 +633,22 @@ proc procTypeRel(c: var TCandidate, f, a: PType): TTypeRelation =
...
+      # X..Y and C..D overlap iff (X <= D and C <= Y)
+      isConvertible
+    else:
+      isNone
+  
+  if f.isOrdinalType: 
+    checkRange(firstOrd(a), lastOrd(a), firstOrd(f), lastOrd(f))
+  else: 
+    check_range(firstFloat(a), lastFloat(a), firstFloat(f), lastFloat(f))
```
Comment (Araq) [inline compiler/sigmatch.nim]: ``checkRange``

File: compiler/sigmatch.nim [hunk 8]
```diff
@@ -633,20 +633,22 @@ proc procTypeRel(c: var TCandidate, f, a: PType): TTypeRelation =
...
+      isEqual
+    elif afirst >= ffirst and alast <= flast:
+      isConvertible
+    elif afirst <= flast and ffirst <= a_last:
+      # X..Y and C..D overlap iff (X <= D and C <= Y)
+      isConvertible
+    else:
+      isNone
+  
+  if f.isOrdinalType:
```
Comment (Araq) [inline compiler/sigmatch.nim]: Move this condition out of ``typeRangeRel``. The callers should know which variant is used, there is no need to check for ``f.isOrdinalType`` at runtime.

File: compiler/sigmatch.nim (hunk 5 above)
Comment (cooldome) [reply compiler/sigmatch.nim]: fixed

File: compiler/sigmatch.nim (hunk 7 above)
Comment (cooldome) [reply compiler/sigmatch.nim]: fixed

File: doc/manual.rst (hunk 4 above)
Comment (cooldome) [reply doc/manual.rst]: fixed

File: compiler/sigmatch.nim (hunk 8 above)
Comment (cooldome) [reply compiler/sigmatch.nim]: Not sure about this one.  Unless I got it completely wrong, ``typeRangeRel`` used only once and it is in ``typeRel``. It doesn't look like ``typeRel`` in tyRange branch knows the range type, so the same if statement will be required in ``typeRel``. Given that ``typeRel`` is huge, I would rather move if statement outside of ``typeRel``.

Let me know what you think.

**Comment by cooldome** [comment]
ready for review

File: compiler/types.nim [hunk 9]
```diff
@@ -624,6 +628,21 @@ proc firstOrd*(conf: ConfigRef; t: PType): BiggestInt =
...
+
+proc firstFloat*(t: PType): BiggestFloat =
+  case t.kind
+  of tyFloat..tyFloat128: -Inf
+  of tyRange:
+    assert(t.n != nil)        # range directly given:
+    assert(t.n.kind == nkRange)
+    getFloatValue(t.n.sons[0])
+  of tyVar: firstFloat(t.sons[0])
+  of tyGenericInst, tyDistinct, tyTypeDesc, tyAlias, tyStatic, tyInferred:
```
Comment (Araq) [inline compiler/types.nim]: At least ``tySink``  is missing in this list.

File: compiler/types.nim [hunk 10]
```diff
@@ -664,6 +683,22 @@ proc lastOrd*(conf: ConfigRef; t: PType; fixedUnsigned = false): BiggestInt =
...
+
+proc lastFloat*(t: PType): BiggestFloat =
+  case t.kind
+  of tyFloat..tyFloat128: Inf
+  of tyVar: lastFloat(t.sons[0])
+  of tyRange:
+    assert(t.n != nil)        # range directly given:
+    assert(t.n.kind == nkRange)
+    getFloatValue(t.n.sons[1])
+  of tyGenericInst, tyDistinct, tyTypeDesc, tyAlias, tyStatic, tyInferred:
```
Comment (Araq) [inline compiler/types.nim]: Same.

**Comment by Yardanico** [comment]
This fixes https://github.com/nim-lang/Nim/issues/1095

**Comment by Yardanico** [comment]
https://github.com/nim-lang/Nim/issues/1095 can be closed after this PR will be merged

File: compiler/types.nim (hunk 9 above)
Comment (cooldome) [reply compiler/types.nim]: updated

File: compiler/types.nim (hunk 10 above)
Comment (cooldome) [reply compiler/types.nim]: updated

**Pull request merged**